### PR TITLE
Replace legacy validate_* calls with puppet4 datatypes

### PR DESCRIPTION
**Pull request by bastelfreak**
So Puppet 3 is end of life since december 2016. Also Puppetlabs deprecated all the validate_* functions in stdlib. This Pull Requests replaces all the legacy code with datatypes.

### CODE REVIEW

**Comment by bastelfreak** [comment]
Hi everybody, could I please get some feedback on the PR? The acceptance tests are failing but this looks more like a jenkins issue?

File: manifests/job.pp [hunk 1]
```diff
@@ -28,23 +28,17 @@
...
-  $template = undef,
-  $jobname  = $title,
-  $enabled  = undef,
-  $ensure   = 'present',
-  $difftool = '/usr/bin/diff -b -q',
+  String $config,
+  Optional[Stdlib::Absolutepath] $source    = undef,
+  Optional[Stdlib::Absolutepath] $template  = undef,
+  String $jobname                           = $title,
+  $enabled                                  = undef,
```
Comment (vStone) [inline manifests/job.pp]: It might be more useful to add `Any` here. Now it's weird that there is nothing.

File: manifests/job/present.pp [hunk 2]
```diff
@@ -19,16 +19,12 @@
...
 define jenkins::job::present(
-  $config      = undef,
-  $config_file = undef,
-  $jobname  = $title,
-  $enabled  = undef,
-  $difftool = '/usr/bin/diff -b -q',
+  Optional[String] $config      = undef,
+  Optional[String] $config_file = undef,
+  String $jobname               = $title,
+  $enabled                      = undef,
```
Comment (vStone) [inline manifests/job/present.pp]: I prefer `Any`

**Comment by bastelfreak** [comment]
hi @vStone, I added Any to all parameters without datatype. Drop legacy parameters is always nice, but that should happen in another PR. My goal here is just to get rid of all the validate_* functions calls as they are deprecated.

**Comment by bastelfreak** [comment]
Hi,
I'm not sure why travis fails, I can't reproduce this locally. Can anybody help me out here?

File: manifests/init.pp [hunk 3]
```diff
@@ -272,85 +272,47 @@
...
+  Optional[String] $service_provider              = $jenkins::params::service_provider,
+  Hash $config_hash                               = {},
+  Hash $plugin_hash                               = {},
+  Hash $job_hash                                  = {},
+  Hash $user_hash                                 = {},
+  Boolean $configure_firewall                     = false,
+  Boolean $install_java                           = $jenkins::params::install_java,
+  Optional[String] $repo_proxy                    = undef,
+  Optional[String] $proxy_host                    = undef,
+  Integer $proxy_port                             = 3218,
```
Comment (alvin-huang) [inline manifests/init.pp]: Why is this defaulted to 3218?

Comment (bastelfreak) [inline manifests/init.pp]: oh that is a typo and should be 3128

**Comment by bastelfreak** [comment]
this PR should be green after https://github.com/jenkinsci/puppet-jenkins/pull/775 gets merged

File: manifests/init.pp [hunk 4]
```diff
@@ -272,85 +272,47 @@
...
+  Optional[String] $service_provider              = $jenkins::params::service_provider,
+  Hash $config_hash                               = {},
+  Hash $plugin_hash                               = {},
+  Hash $job_hash                                  = {},
+  Hash $user_hash                                 = {},
+  Boolean $configure_firewall                     = false,
+  Boolean $install_java                           = $jenkins::params::install_java,
+  Optional[String] $repo_proxy                    = undef,
+  Optional[String] $proxy_host                    = undef,
+  Integer $proxy_port                             = 3128,
```
Comment (alvin-huang) [inline manifests/init.pp]: Should this be set at all though?

Comment (bastelfreak) [inline manifests/init.pp]: ah right, it isn't set in master. I will set it to optional[integer]

**Comment by jhoblitt** [comment]
Rebased on current master as #793

**Pull request closed without merging**